### ISSUE TITLE
Should MojoAsyncWaiter have an AsyncRun()?

### ISSUE THREAD

**Issue by viettrungluu-cr** [domokit/mojo#749]
I.e., post (delayed) task?

Rationale:
* It's a capability that message loops always have anyway.
* It's a capability that MojoAsyncWaiter kind of already has, in a roundabout/inefficient way: you can wait with a timeout on a handle with a condition that you'll never satisfy (but is theoretically satisfiable, obviously).

(Providing a suitable C++ API/wrapper is a different story.)

I'd make it something like:
```
typedef void (*MojoAsyncRunCallback)(void* closure);

struct MojoAsyncWaiter {
  ...

  // [Order of arguments subject to argument. Maybe |delay| should be last.]
  MojoAsyncWaitID (*AsyncRun)(MojoDeadline delay,
                              MojoAsyncRunCallback callback,
                              void* closure);
```
(The callback would be cancellable using the returned ID. Arguably, `MojoAsyncWaitID` and `CancelWait` should be renamed, but \<shrug>.)

@jamesr, @j9brown -- what do you think?

**Comment by j9brown** [comment]
This would definitely be useful, though I would be inclined to promote the notion of scheduling tasks on a run loop to first class instead of bolting it onto MojoAsyncWaiter.

(This is mostly name bikeshedding.  Renaming MojoAsyncWaiter to MojoTaskRunner / MojoRunLoop / MojoLooper or whatever would be sufficient to express the change in scope.)

FWIW, I actually want a version of this that takes an explicit timestamp (that might be in the past, present, or future) instead of a delay and some guarantees about temporal order of execution.

**Comment by viettrungluu-cr** [comment]
On Tue, Apr 12, 2016 at 2:29 PM, Jeff Brown <notifications@github.com>
wrote:

> This would definitely be useful, though I would be inclined to promote the
> notion of scheduling tasks on a run loop to first class instead of bolting
> it onto MojoAsyncWaiter.
>
> (This is mostly name bikeshedding. Renaming MojoAsyncWaiter to
> MojoTaskRunner / MojoRunLoop / MojoLooper or whatever would be sufficient
> to express the change in scope.)
>
We can argue about naming separately. :P

> FWIW, I actually want a version of this that takes an explicit timestamp
> (that might be in the past, present, or future) instead of a delay and some
> guarantees about temporal order of execution.
>
Yeah, this should be a part of deeper discussion about actual deadlines
(fixed points on some clock) versus timeouts (currently misnamed
"deadlines").

Maybe MojoDeadline should be a struct like:

struct MojoDeadline {
  MojoDeadlineType type;
  uint64_t value;  // [Can argue about signedness here.]
};

where |type| is an "enumeration" with things like
"ABSOLUTE_MONOTONIC_CLOCK" or "RELATIVE_MONOTONIC_CLOCK" (maybe etc., for
different clock types). (So "absolute" is a deadline, whereas "relative" is
a timeout.)

Of course, you could just insist on always using absolute deadlines, and
have those who want relative timeouts compute the deadline ... but then you
really have to ensure that MojoGetTimeTicksNow() (or whatever) is
super-fast.

But, again, we should probably discuss this separately.

**Comment by jamesr** [comment]
I feel that this should be orthogonal to MojoAsyncWaiter.  I think it's useful to have the concept of MojoAsyncWaiter be restricted only to the operation of tying handles to something that can service them without assuming that the implementation of that is something like a message loop.  It certainly doesn't have to be today - the Go equivalent maps each pending operation to a channel and lets the Go routine on the other end decide how to mux/select/whatever over the pending ops.  A C++ fiber/green thread based implementation could do something similar.  Requiring that these implementations also support running arbitrary work on a delay would make them considerably more complicated even when this capability is unnecessary.

In cases where we may want to write generic code that has to run over a message loop type construct we should require an additional interface to the thing currently called MojoAsyncWaiter and let the environment figure out what to provide for this (if anything).  I'd be curious what the specific use cases are for things that need to depend on delayed tasks but otherwise depend only on the C bindings level.

**Comment by j9brown** [comment]
Some good use-cases are timeouts or deferring work to run later (outside of
a critical section perhaps).  It's a pretty common need.  Lots of reusable
library code benefits from having indirect access to such primitives (since
in general someone else is driving the loop and they don't / can't know
who).  A few things in mojo/ui and mojo/services/media are like that today.

Anyhow, I agree this functionality can be layered though some clients may
care about their interactions to a degree (which may be particular to the
environment).

**Comment by jamesr** [comment]
A library needing such functionality is probably written on some higher layers (like something higher level than the raw C bindings layer).  The mojo/ui and mojo/services/media libraries today depend on C++ and many other libs.  Maybe the more useful question to explore is what dependencies should high-level libraries like these should have and what assumptions those dependencies can make.

**Comment by viettrungluu-cr** [comment]
I guess we've conflated two uses of MojoAsyncWaiter:

1) The reason for its existence is that generic (system-provided)
libraries, like GL, need the async-waiting facility. I don't know if they
need/should ever need a more generic task-running facility. (As I said, *a*
reason for saying that MojoAsyncWaiter should have such an explicit
facility is that it already has it implicitly, in an inefficient,
roundabout way.)

2) We also use it in higher-level (C++) libraries (including bindings, and
other C++ libraries). But possibly the C++ Environment should provide a
higher-level thing with more capabilities. (There may be other reasons for
doing this, but I don't want to get into that here.)

**Comment by jamesr** [comment]
Right, I'm talking about (1).  I'm reluctant to require that implementations of (1) provide this functionality in all cases, i.e. that you would need to have a timed task runner implementation in order to use the GL bindings.  How about we add this as a separate piece of functionality, see how the use cases play out, and then if it turns out it's always needed we can choose to combine with the existing MojoAsyncWaiter later on.

**Comment by j9brown** [comment]
Works for me.  :)
